contributivity_measures.py

```diff
@@ -168,3 +166,704 @@ def not_twice_characteristic(permut):
         "prop": sv / np.sum(sv),
         "computed_val": char_value_dict,
     }
+
+
+# %% compute Shapley values with the truncated Monte-carlo method with a small bias correction
+
+
+def interpol_trunc_MC(the_scenario, sv_accuracy=0.01, alpha=0.9, truncation=0.05):
+    """Return the vector of approximated Shapley value corresponding to a list of node and a characteristic function using the truncated monte-carlo method with a small correction of the bias"""
+
+    preprocessed_node_list = the_scenario.node_list
+    n = len(preprocessed_node_list)
+
+    # We store the value of the characteristic function in order to avoid recomputing it twice
+    char_value_dict = {
+        (): 0
+    }  # the dictionary that will countain the values #here we suppose the characteristic function is 0 for the empty set
+
+    # Return the characteristic function of the nodelist associated to the ensemble permut, without recomputing it if it was already computed
+    def not_twice_characteristic(permut):
+        # Sort permut
+        permut = np.sort(permut)
+        try:  # Return the characteristic_func(permut) if it was already computed
+            return char_value_dict[tuple(permut)]
+        except KeyError:  # Characteristic_func(permut) has not been computed yet, so we compute, store, and return characteristic_func(permut)
+            small_node_list = np.array([preprocessed_node_list[i] for i in permut])
+            char_value_dict[tuple(permut)] = fl_training.compute_test_score(
+                small_node_list,
+                the_scenario.epoch_count,
+                the_scenario.x_esval,
+                the_scenario.y_esval,
+                the_scenario.x_test,
+                the_scenario.y_test,
+                the_scenario.is_early_stopping,
+                save_folder=the_scenario.save_folder,
+            )
+
+            return char_value_dict[tuple(permut)]
+
+    characteristic_all_node = not_twice_characteristic(
+        np.arange(n)
+    )  # Characteristic function on all nodes
+    if n == 1:
+        return {
+            "sv": characteristic_all_node,
+            "std_sv": np.array([0]),
+            "prop": np.array([1]),
+            "computed_val": char_value_dict,
+        }
+    else:
+        contributions = np.array([[]])
+        permutation = np.zeros(n)  # Store the current permutation
+        t = 0
+        q = -norm.ppf((1 - alpha) / 2, loc=0, scale=1)
+        v_max = 0
+        while (
+            t < 100 or t < 4 * q ** 2 * v_max / (sv_accuracy) ** 2
+        ):  # Check if the length of the confidence interval  is below the value of sv_accuracy*characteristic_all_node
+            t += 1
+            print(t)
```

Comment:
The prinds floods the output. We can remove the print or print every 1000th iteration for instance ?